File: LC/python/Permutations II.py
class Solution:
    # @param num, a list of integer
    # @return a list of lists of integers
    def permuteUnique(self, num):
        if len(num) == 0:
            return []
        elif len(num) == 1:
            return [num]
        res = []
        num.sort()
        pre = 99999
        for i in range(len(num)):
            if pre == num[i]:
                continue
            pre = num[i]
            for j in self.permuteUnique(num[0:i]+num[i+1::]):
                    res.append([num[i]]+j)
        return res

File: LC/java/Permutations II.java
public class Solution {
    public ArrayList<ArrayList<Integer>> permuteUnique(int[] num) {
        ArrayList<ArrayList<Integer>> res = new ArrayList<ArrayList<Integer>>();
        if(num.length==0){
            return res;
        }
        
        int[] visited = new int[num.length];
        ArrayList<Integer> sub = new ArrayList<Integer>();
        Arrays.sort(num);
        dfs(res,sub,num,visited);
        return res;
    }
    
    public void dfs(ArrayList<ArrayList<Integer>> res, ArrayList<Integer>sub, int[] num, int[] visited){
        if (sub.size()==num.length){
            res.add(new ArrayList<Integer>(sub));
            return;
        }
        
        for(int i = 0 ; i < num.length; i++){
            if(visited[i] == 1 || (i!=0 && num[i]==num[i-1] && visited[i-1]==0) ) continue;
            
            visited[i] = 1;
            sub.add(num[i]);
            dfs(res,sub,num,visited);
            sub.remove(sub.size()-1);
            visited[i] = 0;
            
        }
    }
}